File: src/algorithms/hash-maps/two-sum/sources/two-sum.py
# Two Sum — find two indices whose values add up to the target using a hash map
def two_sum(numbers, target):
    map = {}  # @step:initialize
    for idx, num in enumerate(numbers):
        complement = target - num  # @step:lookup-key
        if complement in map:  # @step:key-found
            return [map[complement], idx]  # @step:key-found
        # Complement not found — store current number for future lookups
        map[num] = idx  # @step:insert-key
    return [-1, -1]  # @step:complete

File: src/algorithms/hash-maps/lookup/two-sum/sources/two-sum.go
// Two Sum — find two indices whose values add up to the target using a hash map
package main

func twoSum(numbers []int, target int) [2]int {
	numMap := make(map[int]int) // @step:initialize
	for idx, current := range numbers {
		complement := target - current // @step:lookup-key
		if storedIdx, exists := numMap[complement]; exists {
			// @step:key-found
			return [2]int{storedIdx, idx} // @step:key-found
		}
		// Complement not found — store current number for future lookups
		numMap[current] = idx // @step:insert-key
	}
	return [2]int{-1, -1} // @step:complete
}
